docs/proposals/component-registration.md

```diff
@@ -0,0 +1,406 @@
+
+## Terminology
+
+A Kubernetes Cluster **Component**...
+
+- Is considered a part of the cluster
+- Must register (or be registered) with Kubernetes (apiserver) to join the cluster
+- Must know about Kubernetes (speak the API)
+- Must be deployed/managed by the cluster admin/operator (rather than a user)
+- Should expose a method for remotely probing the liveness and readiness of each instance (e.g. tcp/http/https).
+- May be scheduled/deployed on the cluster (primary components vs addon components)
+- May be required for the cluster to function (required vs optional)
+- May be added/removed from a running cluster (mutable cluster component membership over time)
+- May require admin API access (privileged behavior)
+- May consist of one or more instances
+- May be deployed on different (network accessible) machines
+- May be upgraded over time
+
+The **External Components** are components that are deployed and lifecycle managed by some external system.
+
+The **Addon Components** are components that are deployed and lifecycle managed by the Kubernetes to which they provide functionality. These components must be optional in order for them to be deployable by Kubernetes (at least until an HA bootstrapping method is developed). Kubernetes conformance tests currently require some of these (e.g. kube-dns), but they are not required for orchestration and scheduling functionality.
+
+The **Primary Components** (currently apiserver, controller-manager, and scheduler) are components that are required for orchestration and scheduling functionality. These components are currently all External Components but may become Addon Components, once an HA bootstrapping method is developed.
+
+The"**Hyperkube** is a single binary that can be run as each of the three primary component binaries. Hyperkube may also refer to the set (or pod) of primary component containers or the (virtual) machine that those containers run on. Not all Kubernetes deployments use or should be required to use the Hyperkube.
+
+**Master** has been sometimes used in reference to both the APIServer component and the (virtual) machine which hosts the three primary components. I'm going to avoid this term, because it is ambiguous.
+
+**Etcd** is a consistent key value storage cluster. It should not be considered a component, because it is a dependency of the apiserver, does not know about Kubernetes, and thus cannot register itself with Kubernetes but must be configured. In general, the apiserver attempts to hide/abstract Etcd from the rest of the components with its API.
+
+**ComponentStatuses** is a top-level Kubernetes API resource that exposes a `/componentstatuses` endpoint that shows component and Etcd status. It's unquestionably useful and used, but questionably named and implemented. However, it must continue to function (until v2) for reverse compatibility. See [Problems](#problems) for more details. `kubectl get componentstatuses` can be used from the CLI as a wrapper for the API..
```

Comment:
Right, I want aggregate-of-aggregate checking; suppose you're doing the
microservices thing (k8s is an example) and you want to know if your
complete stack is up: I want a thing to check if all of your microservices
are up.

So e.g. you'd have scheduler, controller-manager, and apiserver services
(either self-hosted or externally hosted, doesn't matter); and your entire
cluster stack is healthy if all of those services are healthy.

On Mon, Oct 12, 2015 at 8:26 AM, Karl Isenberg <notifications@github.com>
wrote:

> In docs/proposals/component-registration.md
> <https://github.com/kubernetes/kubernetes/pull/13216#discussion_r41768148>
> :
>
> > +- May be deployed on different (network accessible) machines
> > +- May be upgraded over time
> > +
> > +The **External Components** are components that are deployed and lifecycle managed by some external system.
> > +
> > +The **Addon Components** are components that are deployed and lifecycle managed by the Kubernetes to which they provide functionality. These components must be optional in order for them to be deployable by Kubernetes (at least until an HA bootstrapping method is developed). Kubernetes conformance tests currently require some of these (e.g. kube-dns), but they are not required for orchestration and scheduling functionality.
> > +
> > +The **Primary Components** (currently apiserver, controller-manager, and scheduler) are components that are required for orchestration and scheduling functionality. These components are currently all External Components but may become Addon Components, once an HA bootstrapping method is developed.
> > +
> > +The"**Hyperkube** is a single binary that can be run as each of the three primary component binaries. Hyperkube may also refer to the set (or pod) of primary component containers or the (virtual) machine that those containers run on. Not all Kubernetes deployments use or should be required to use the Hyperkube.
> > +
> > +**Master** has been sometimes used in reference to both the APIServer component and the (virtual) machine which hosts the three primary components. I'm going to avoid this term, because it is ambiguous.
> > +
> > +**Etcd** is a consistent key value storage cluster. It should not be considered a component, because it is a dependency of the apiserver, does not know about Kubernetes, and thus cannot register itself with Kubernetes but must be configured. In general, the apiserver attempts to hide/abstract Etcd from the rest of the components with its API.
> > +
> > +**ComponentStatuses** is a top-level Kubernetes API resource that exposes a `/componentstatuses` endpoint that shows component and Etcd status. It's unquestionably useful and used, but questionably named and implemented. However, it must continue to function (until v2) for reverse compatibility. See [Problems](#problems) for more details. `kubectl get componentstatuses` can be used from the CLI as a wrapper for the API..
>
> @lavalamp <https://github.com/lavalamp>
> The current proposal will allow these approaches:
>
>    1. Query all the component specs:
>       - Pods w/ namespace=kube-system & type=component & component=<name>
>       - ExternalServiceProviders w/ namespace=kube-system & type=component
>       & component=<name>
>
>
>    1. Query individual component status:
>    2. Pods w/ namespace=kube-system & type=component & component=<name>
>    3.
>
>    ExternalServiceProviders w/ namespace=kube-system & type=component &
>    component=<name>
>    4.
>
>    Query aggregate Service status:
>    5.
>
>    Services w/ namespace=kube-system & type=component & component=<name>
>    6.
>
>    Query individual component readiness (only-readiness, not all
>    conditions):
>    7. Endpoints w/ namespace=kube-system & type=component &
>    component=<name>
>
> Unfortunately Endpoints isn't structured such that we can add other status
> conditions, otherwise they'd be there. But it sounds like that's similar to
> what you're asking for. However, the way you proposed it was with
> ServiceStatus, which is an aggregate, not an individual status. I'm not
> sure why you would need a group of aggregates, but you could get it with
> option 3 without specifying a component name, or by using some other group
> label.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13216/files#r41768148>.
>